### PR TITLE
fix: delete federated resources prior to uninstallation

### PR DESCRIPTION
**Pull request by makkes**
**What type of PR is this?**

Bug

**What this PR does/ why we need it**:

All the federated resources are now deleted as part of a pre-delete
hook so that the kubefed controller has the chance to remove the
finalizer from those resources. Without this change, all the federated
resources have always been left behind on clusters when the chart was
deleted because they would have the dangling finalizers on them.

The reason for this is that the controller managing these finalizers
(kubefed controller) is removed before the resources themselves.

**Which issue(s) this PR fixes**:

https://jira.d2iq.com/browse/D2IQ-72106

**Special notes for your reviewer**:
I manually tested installation and uninstallation of this chart. After it has been deleted, no federated types are left on the cluster.

**Does this PR introduce a user-facing change?**:

```release-note
All federated resources are now deleted as soon as Kommander is uninstalled.
```

**Checklist**

* [x] *If a chart is changed, the chart version is correctly incremented.*
* [x] The commit message explains the changes and why are needed.
* [x] The code builds and passes lint/style checks locally.
* [ ] The relevant subset of integration tests pass locally.
* [ ] The core changes are covered by tests.
* [ ] The documentation is updated where needed.

### CODE REVIEW

File: stable/kommander/templates/pre-deletion-purger.yaml
Comment (jimmidyson) [inline stable/kommander/templates/pre-deletion-purger.yaml:14]:
```suggestion
      kubectl delete ${ft} --all --all-namespaces
```

File: stable/kommander/templates/pre-deletion-purger.yaml
Comment (jimmidyson) [inline stable/kommander/templates/pre-deletion-purger.yaml:79]: Let's use a specific version, say `v1.19.3`?

File: stable/kommander/templates/pre-deletion-purger.yaml
Comment (jimmidyson) [inline stable/kommander/templates/pre-deletion-purger.yaml:28]: Could we limit the scope here to the specific federated types API group?

File: stable/kommander/templates/pre-deletion-purger.yaml
Comment (jimmidyson) [inline stable/kommander/templates/pre-deletion-purger.yaml:76]:
```suggestion
      serviceAccountName: purger
```

**Comment by jimmidyson** [comment]
I thought we said that we would target the specific resources installed as part of the chart via label selection in the first cut, and provide the full purge as optional?

File: stable/kommander/templates/pre-deletion-purger.yaml
Comment (makkes) [reply stable/kommander/templates/pre-deletion-purger.yaml:28]: I explicitly made this a wildcard because if anyone adds a resource from an additional API group, this job would fail.

**Comment by makkes** [comment]
> I thought we said that we would target the specific resources installed as part of the chart via label selection in the first cut, and provide the full purge as optional?

You are right. One problem that I encountered is that Helm doesn't put labels but annotations on the resources so I'd have to put more filtering logic into the script. Do we want to do that? Honest question.

**Comment by jimmidyson** [comment]
We add labels to all other resources via the chart... https://github.com/mesosphere/yakcl/blob/master/chart/kommander-federation/templates/issuer_utility_apiserver.yaml#L6-L7... but not federated* resources deployed via the chart :facepalm:

File: stable/kommander/templates/pre-deletion-purger.yaml
Comment (hectorj2f) [inline stable/kommander/templates/pre-deletion-purger.yaml:13]: What about the federated resources created in namespace kommander-system?

Comment (hectorj2f) [reply stable/kommander/templates/pre-deletion-purger.yaml:13]: Do we need to delete them?

Comment (makkes) [reply stable/kommander/templates/pre-deletion-purger.yaml:13]: This line only searches for `FederatedTypeConfigs` and the Kommander chart does not create any `FederatedTypeConfigs` outside of the `kommander` namespace.

Comment (hectorj2f) [reply stable/kommander/templates/pre-deletion-purger.yaml:13]: True!

**Review by hectorj2f — APPROVED** [review]

**Review by armandgrillet — APPROVED** [review]

**Comment by makkes** [comment]
Will wait for @jimmidyson's final approval.

**Review by jimmidyson — CHANGES_REQUESTED** [review]
Chart version needs a bump but otherwise :ok_hand:

**Review by jimmidyson — APPROVED** [review]

**Pull request merged**
